sympy/functions/combinatorial/numbers.py

```diff
@@ -910,15 +911,13 @@ def eval(cls, n, m=None):
                 return S.Infinity
             elif is_gt(m, S.One):
                 return zeta(m)
+        elif m.is_Integer and m.is_nonpositive:
+            return (bernoulli(1-m, n+1) - bernoulli(1-m)) / (1-m)
         elif n.is_Integer:
-            if n.is_negative:
-                return S.NaN
-            elif m not in cls._functions:
-                @recurrence_memo([0])
-                def f(n, prev):
-                    return prev[-1] + S.One / n**m
-                cls._functions[m] = f
-            return cls._functions[m](int(n))
+            if n.is_negative and (m.is_integer is False or m.is_nonpositive is False):
+                return S.ComplexInfinity if m is S.One else S.NaN
+            if n.is_nonnegative:
+                return sum(k**(-m) for k in range(1, int(n)+1))
```

Comment:
Using `sum` with SymPy expressions tends to work out as `O(N**2)` because `Add.flatten` which is `O(N)` in the number of terms is called for each term. It's more efficient to use `Add(*terms)`:
```python
return Add(*(k**(-m) for k in range(1, int(n)+1)))
```

Reply:
You can time this with something like `harmonic(10000, x)`.

Reply:
Wait until I commit this…

Reply:
OK, you may merge now!